modulus/sym/amp.py

```diff
@@ -0,0 +1,634 @@
+
+    def unscale_(self):
+        raise RuntimeError("Function should not be called")
+
+    def step(self):
+        raise RuntimeError("Function should not be called")
+
+    def scale(self, *args, **kwargs):
+        out = super().scale(*args, **kwargs)
+        if not self._enabled:
+            return out
+        if self._found_inf is None:
+            self._lazy_init_found_inf()
+        return out
+
+    def unscale_deriv(self, grad: List[Tensor]):
+        """
+        Return unscaled derivatives which are also checked for infs/NaNs.
+
+        Parameters
+        ----------
+        grad : List[Tensor]
+            A list of derivative tensors need to be unscaled.
+
+        Returns
+        -------
+        List[Tensor]
+            Unscaled derivatives
+        """
+        if not self._enabled:
+            return grad
+
+        self._check_scale_growth_tracker("unscale_deriv")
+        self._check_found_inf("unscale_deriv")
+
+        inv_scale = self._scale.reciprocal()
+        found_inf = torch.full((1,), 0.0, dtype=torch.float32, device=inv_scale.device)
+        grad = foreach_non_finite_check_and_unscale(found_inf, inv_scale, grad)
+
+        self._found_inf += found_inf
+        return grad
+
+    @property
+    def found_inf(self):
+        """
+        Returns a Python bool containing the current found_inf state.
+        If the scaler is disabled or there are no derivative nodes in the model, return False.
+
+        .. warning::
+            :meth:`found_inf` incurs a CPU-GPU sync.
+
+        .. warning::
+            :meth:`found_inf` should be called before function `update`, since `update` will reset the
+            _found_inf state.
+        """
+        if not self._enabled:
+            return False
+
+        if self._found_inf == None:
+            logger.warning(
+                "deriv_scaler.scale() never got called, "
+                "please make sure there are no derivative nodes in the model."
+            )
+            return False
+
+        _found_inf = self._check_found_inf("found_inf")
+        return _found_inf.bool().item()
+
+    def _get_found_inf_tensor(self):
+        """
+        Return the tensor of _found_inf, should only be called when scaler is enabled.
+        """
+        _found_inf = self._check_found_inf("_get_found_inf_tensor")
+        return _found_inf
+
+    def update(self):
+        """
+        Updates the scale factor.
+
+        If self._found_inf > 0, the scale is multiplied by ``backoff_factor`` to reduce it.
+        If ``growth_interval`` unskipped iterations occurred consecutively, the scale is
+        multiplied by ``growth_factor`` to increase it.
+        If the scaling factor is less or equal to the recover_threshold, it will grow more
+        frequently (using the recover_growth_interval) to avoid decreasing too small and
+        stopping the training process.
+        The max_scale defins a upper bound for the scaling factor.
```

Comment:
```suggestion
        If self._found_inf > 0, the scale is multiplied by ``backoff_factor`` to decrease it.
        If there are ``growth_interval`` consecutive unskipped iterations, the scale is
        multiplied by ``growth_factor`` to increase it.
        If the scaling factor is less or equal to the recover_threshold, it will grow more
        frequently (using the recover_growth_interval) to avoid decreasing too small and
        stopping the training process.
        The max_scale defines an upper bound for the scaling factor.
```

Reply:
Done.